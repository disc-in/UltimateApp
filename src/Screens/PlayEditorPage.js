import React, { useState, useEffect, useLayoutEffect } from 'react';
import { StyleSheet, View, Platform } from 'react-native';
import { MaterialCommunityIcons } from '@expo/vector-icons';
import { connect } from 'react-redux';

import I18n from '../utils/i18n';
import theme from '../styles/theme.style';
import { savePlay } from '../Store/Actions/playAction';
import Drill from '../Components/animation/Drill';
import PlayTitle from '../Components/editor/PlayTitle';
import AnimationEditor from '../Components/editor/AnimationEditor';
import AnimationHistory from '../Components/editor/toolbar/AnimationHistory';
import SharePlay from '../Components/editor/toolbar/SharePlay';

export const PlayEditorPage = (props) => {
  const { navigation, route } = props;
  const [currentPlay, setCurrentPlay] = useState({
    ...route.params.currentPlay,
    animation: new Drill(route.params.currentPlay.animation),
  });

  // isPlaySaved is true unless there are unsaved changes on the current play
  const [isPlaySaved, setIsPlaySaved] = useState(true);

  useLayoutEffect(() => {
    navigation.setOptions({
      headerTitleContainerStyle: {
        ...Platform.select({
          ios: {
            marginLeft: 40,
            marginRight: 0,
            flexGrow: 1,
          },
        }),
      },
      headerTitle: () => <PlayTitle play={currentPlay} safe unsavedPlay={!isPlaySaved} onPress={() => {}} />,
    }),
      [];
  });

  const onAnimationChange = (animation) => {
    setIsPlaySaved(false);
    const newPlay = { ...currentPlay, animation };
    setCurrentPlay(newPlay);
    props.savePlay(newPlay);
    setTimeout(() => setIsPlaySaved(true), 1000);
  };

  return (
    <View style={styles.playEditorPage}>
      <View style={styles.centeredView}>
        <AnimationEditor
          onAnimationChange={onAnimationChange}
          animation={currentPlay.animation}
          uuid={currentPlay.uuid}
        />
      </View>
      <View style={styles.toolBar}>
        <View style={styles.toolBarItem}>
          <AnimationHistory animation={currentPlay.animation} onAnimationHistoryChange={onAnimationChange} />
        </View>
        <View style={styles.separator} />
        <View style={styles.toolBarItem}>
          <SharePlay currentPlay={currentPlay} />
        </View>
      </View>
    </View>
  );
};

const mapStateToProps = (state) => {
  return {
    customPlays: state.customPlays,
  };
};

const mapDispatchToProps = { savePlay };

export default connect(mapStateToProps, mapDispatchToProps)(PlayEditorPage);

const styles = StyleSheet.create({
  playEditorPage: {
    flex: 1,
  },
  centeredView: {
    backgroundColor: theme.BACKGROUND_COLOR_LIGHT,
    flex: 1,
  },
  toolBar: {
    height: '8%',
    minHeight: 40,
    width: '100%',
    backgroundColor: theme.COLOR_PRIMARY,
    position: 'absolute',
    bottom: 0,
    alignItems: 'center',
    flexDirection: 'row',
    justifyContent: 'space-evenly',
  },
  toolBarItem: {
    flex: 1,
    alignItems: 'center',
    paddingHorizontal: 10,
  },
  separator: {
    height: '70%',
    borderRightWidth: 0.5,
    borderRightColor: theme.COLOR_SECONDARY,
  },
});
