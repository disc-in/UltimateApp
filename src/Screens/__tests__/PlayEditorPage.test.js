import React from 'react';
import { render } from '@testing-library/react-native';
import { Provider } from 'react-redux';

import store from '../../Store/testStore';
import Drill from '../../Components/animation/Drill';

import PlayEditorPage from '../PlayEditorPage';

jest.mock('react-native-reanimated', () => require('react-native-reanimated/mock'));

describe('<PlayEditorPage />', () => {
  it('renders correctly', async () => {
    const navigation = { navigate: jest.fn(), setOptions: jest.fn() };
    const play = {
      uuid: undefined,
      animation: new Drill(),
      title: 'title',
    };
    const route = {
      params: {
        currentPlay: play,
      },
    };

    const { toJSON } = render(
      <Provider store={store}>
        <PlayEditorPage navigation={navigation} route={route} />
      </Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
