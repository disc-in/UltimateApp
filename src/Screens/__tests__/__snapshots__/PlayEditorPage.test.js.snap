// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlayEditorPage /> renders correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#fff",
        "flex": 1,
      }
    }
  >
    <View
      onLayout={[Function]}
    >
      <View>
        <View
          onLayout={[Function]}
          style={
            {
              "height": 833.75,
              "width": 750,
            }
          }
        >
          <View
            style={
              {
                "height": 833.75,
                "width": 750,
              }
            }
          >
            <View
              style={
                [
                  {
                    "backgroundColor": "#232b2b",
                    "borderWidth": 1,
                    "position": "absolute",
                  },
                  {
                    "height": 708.6875,
                    "left": 712.5,
                    "top": 41.6875,
                    "width": 3.375,
                  },
                ]
              }
            />
            <View
              style={
                [
                  {
                    "backgroundColor": "#232b2b",
                    "borderWidth": 1,
                    "position": "absolute",
                  },
                  {
                    "height": 708.6875,
                    "left": 37.5,
                    "top": 41.6875,
                    "width": 3.375,
                  },
                ]
              }
            />
            <View
              style={
                [
                  {
                    "backgroundColor": "#232b2b",
                    "borderWidth": 1,
                    "position": "absolute",
                  },
                  {
                    "height": 3.375,
                    "left": 37.5,
                    "top": 41.6875,
                    "width": 675,
                  },
                ]
              }
            />
            <View
              style={
                [
                  {
                    "backgroundColor": "#232b2b",
                    "borderWidth": 1,
                    "position": "absolute",
                  },
                  {
                    "height": 3.375,
                    "left": 37.5,
                    "top": 325.1625,
                    "width": 675,
                  },
                ]
              }
            />
            <Text />
          </View>
          <View
            height="100%"
            style={
              {
                "left": 0,
                "position": "absolute",
                "top": 0,
              }
            }
            width="100%"
          />
        </View>
        <View>
          <View
            style={
              {
                "flexDirection": "row",
                "paddingHorizontal": 10,
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "marginHorizontal": 10,
                  "opacity": 1,
                }
              }
            >
              <Text />
            </View>
            <View
              style={
                {
                  "flex": 1,
                  "marginVertical": 10,
                }
              }
            >
              <View
                style={
                  [
                    {
                      "backgroundColor": "gray",
                      "height": 5,
                      "position": "absolute",
                      "top": 10,
                    },
                    {
                      "width": 307.5,
                    },
                  ]
                }
              />
              <View
                collapsable={false}
                style={
                  {
                    "backgroundColor": "black",
                    "height": 5,
                    "position": "absolute",
                    "top": 10,
                    "width": 0,
                  }
                }
              />
              <View
                style={
                  [
                    {
                      "backgroundColor": "gray",
                      "borderRadius": 15,
                      "height": 15,
                      "position": "absolute",
                      "top": 5,
                      "width": 15,
                    },
                    {
                      "left": 0,
                    },
                  ]
                }
              />
              <View
                style={
                  [
                    {
                      "backgroundColor": "gray",
                      "borderRadius": 15,
                      "height": 15,
                      "position": "absolute",
                      "top": 5,
                      "width": 15,
                    },
                    {
                      "left": 307.5,
                    },
                  ]
                }
              />
              <View
                collapsable={false}
                style={
                  {
                    "backgroundColor": "black",
                    "borderRadius": 15,
                    "height": 15,
                    "left": 0,
                    "opacity": 1,
                    "position": "absolute",
                    "top": 5,
                    "width": 15,
                  }
                }
              />
              <View
                collapsable={false}
                style={
                  {
                    "backgroundColor": "black",
                    "borderRadius": 15,
                    "height": 15,
                    "left": 307.5,
                    "opacity": 0,
                    "position": "absolute",
                    "top": 5,
                    "width": 15,
                  }
                }
              />
              <Text
                style={
                  [
                    {
                      "position": "absolute",
                      "top": 25,
                    },
                    {
                      "left": 3,
                    },
                  ]
                }
              >
                1
              </Text>
              <Text
                style={
                  [
                    {
                      "position": "absolute",
                      "top": 25,
                    },
                    {
                      "left": 310.5,
                    },
                  ]
                }
              >
                2
              </Text>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "height": "100%",
                    "left": -7.5,
                    "opacity": 1,
                    "position": "absolute",
                    "width": 30,
                  }
                }
              />
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "height": "100%",
                    "left": 300,
                    "opacity": 1,
                    "position": "absolute",
                    "width": 30,
                  }
                }
              />
            </View>
            <View
              style={
                {
                  "marginHorizontal": 10,
                }
              }
            >
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "opacity": 1,
                  }
                }
              >
                <Text />
              </View>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "opacity": 1,
                  }
                }
              >
                <Text />
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "height": "12%",
            "marginHorizontal": 10,
          }
        }
      >
        <View
          style={
            [
              {
                "flex": 1,
                "flexDirection": "row",
              },
              {
                "alignItems": "center",
              },
            ]
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "height": 30,
                "justifyContent": "space-around",
                "position": "absolute",
                "width": "100%",
                "zIndex": 1,
              }
            }
          >
            <View
              style={
                {
                  "height": 80,
                  "width": "45%",
                }
              }
            >
              <RNCPicker
                items={
                  [
                    {
                      "label": "Rectangle",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": "rectangle",
                    },
                    {
                      "label": "Endzone",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": "endzone",
                    },
                    {
                      "label": "3/4 field",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": "three quarters field",
                    },
                    {
                      "label": "Empty",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": "empty",
                    },
                  ]
                }
                numberOfLines={1}
                onChange={[Function]}
                selectedIndex={1}
                style={
                  [
                    {
                      "height": 216,
                    },
                    {
                      "fontSize": 14,
                      "height": 80,
                    },
                  ]
                }
              />
            </View>
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "space-evenly",
                  "marginLeft": 20,
                  "width": "45%",
                }
              }
            >
              <View
                collapsable={false}
                onMoveShouldSetResponder={[Function]}
                onMoveShouldSetResponderCapture={[Function]}
                onResponderEnd={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderReject={[Function]}
                onResponderRelease={[Function]}
                onResponderStart={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                onStartShouldSetResponderCapture={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "flexBasis": "25%",
                    "height": "100%",
                    "justifyContent": "center",
                    "minHeight": 0,
                    "minWidth": 0,
                    "padding": 5,
                    "transform": [
                      {
                        "translateX": 0,
                      },
                      {
                        "translateY": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  style={
                    [
                      {
                        "alignItems": "center",
                        "justifyContent": "center",
                        "textAlign": "center",
                        "textAlignVertical": "center",
                      },
                      {
                        "backgroundColor": "#2a9e91",
                      },
                      {
                        "borderRadius": 0,
                        "height": 0,
                        "width": 0,
                      },
                    ]
                  }
                >
                  <Text
                    style={
                      {
                        "color": "#fff",
                        "fontWeight": "bold",
                      }
                    }
                  >
                    1
                  </Text>
                </View>
              </View>
              <View
                collapsable={false}
                onMoveShouldSetResponder={[Function]}
                onMoveShouldSetResponderCapture={[Function]}
                onResponderEnd={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderReject={[Function]}
                onResponderRelease={[Function]}
                onResponderStart={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                onStartShouldSetResponderCapture={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "flexBasis": "25%",
                    "height": "100%",
                    "justifyContent": "center",
                    "minHeight": 0,
                    "minWidth": 0,
                    "padding": 5,
                    "transform": [
                      {
                        "translateX": 0,
                      },
                      {
                        "translateY": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  style={
                    [
                      {
                        "alignItems": "center",
                        "justifyContent": "center",
                        "textAlign": "center",
                        "textAlignVertical": "center",
                      },
                      {
                        "backgroundColor": "#000000",
                      },
                      {
                        "borderRadius": 0,
                        "height": 0,
                        "width": 0,
                      },
                    ]
                  }
                >
                  <Text
                    style={
                      {
                        "color": "#fff",
                        "fontWeight": "bold",
                      }
                    }
                  >
                    1
                  </Text>
                </View>
              </View>
              <View
                collapsable={false}
                onMoveShouldSetResponder={[Function]}
                onMoveShouldSetResponderCapture={[Function]}
                onResponderEnd={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderReject={[Function]}
                onResponderRelease={[Function]}
                onResponderStart={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                onStartShouldSetResponderCapture={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "flexBasis": "25%",
                    "height": "100%",
                    "justifyContent": "center",
                    "minHeight": 0,
                    "minWidth": 0,
                    "padding": 5,
                    "transform": [
                      {
                        "translateX": 0,
                      },
                      {
                        "translateY": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  style={
                    [
                      {
                        "alignItems": "center",
                        "backgroundColor": "white",
                        "borderColor": "black",
                        "justifyContent": "center",
                        "textAlign": "center",
                        "textAlignVertical": "center",
                      },
                      {
                        "borderRadius": 0,
                        "borderWidth": 0,
                        "height": 0,
                        "width": 0,
                      },
                    ]
                  }
                />
              </View>
              <View
                collapsable={false}
                onMoveShouldSetResponder={[Function]}
                onMoveShouldSetResponderCapture={[Function]}
                onResponderEnd={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderReject={[Function]}
                onResponderRelease={[Function]}
                onResponderStart={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                onStartShouldSetResponderCapture={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "flexBasis": "25%",
                    "height": "100%",
                    "justifyContent": "center",
                    "minHeight": 0,
                    "minWidth": 0,
                    "padding": 5,
                    "transform": [
                      {
                        "translateX": 0,
                      },
                      {
                        "translateY": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  style={
                    [
                      {
                        "alignItems": "center",
                        "backgroundColor": "transparent",
                        "borderBottomColor": "#feae19",
                        "borderColor": "transparent",
                        "borderStyle": "solid",
                        "justifyContent": "center",
                        "textAlign": "center",
                        "textAlignVertical": "center",
                      },
                      {
                        "borderBottomWidth": 0,
                        "borderLeftWidth": 0,
                        "borderRightWidth": 0,
                      },
                    ]
                  }
                >
                  <Text
                    style={
                      {
                        "height": 0,
                        "width": 0,
                      }
                    }
                  >
                    1
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#232b2b",
        "bottom": 0,
        "flexDirection": "row",
        "height": "8%",
        "justifyContent": "space-evenly",
        "minHeight": 40,
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "paddingHorizontal": 10,
        }
      }
    >
      <View
        style={
          {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": true,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "flexGrow": 1,
              "justifyContent": "center",
              "opacity": 1,
            }
          }
          testID="undoButton"
        >
          <Text />
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": true,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "flexGrow": 1,
              "justifyContent": "center",
              "opacity": 1,
            }
          }
          testID="redoButton"
        >
          <Text />
        </View>
      </View>
    </View>
    <View
      style={
        {
          "borderRightColor": "#A6A6A6",
          "borderRightWidth": 0.5,
          "height": "70%",
        }
      }
    />
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "paddingHorizontal": 10,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="shareButton"
      >
        <Text />
      </View>
    </View>
  </View>
</View>
`;
